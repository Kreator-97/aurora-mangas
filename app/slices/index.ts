export { default as uiSlice } from './uiSlice'
export { default as authSlice } from './authSlice'
export { default as shoppingCartSlice } from './shoppingCartSlice'
