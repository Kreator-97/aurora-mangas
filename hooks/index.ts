export * from './useForm'
export * from './useCounter'
