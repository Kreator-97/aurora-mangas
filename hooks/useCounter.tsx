import { useState } from 'react'

interface Props {
  initial?: number;
  minValue?: number;
}

export const useCounter = ({initial = 0, minValue }:Props) => {
  const [ counter, setCounter] = useState(initial)

  const increment = (factor:number) => {
    setCounter(counter + factor)
  }

  const decrement = (factor:number) => {
    if ( minValue !== undefined ) {
      const max = Math.max(minValue, counter-factor)
      setCounter(max)
      return
    }
    setCounter(counter - factor)
  }

  const reset = () => setCounter(initial)

  return { increment, decrement, reset, counter }
}
