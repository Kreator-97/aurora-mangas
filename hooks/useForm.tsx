import { ChangeEvent, useState } from 'react'

export function useForm<T> (initialValue: T) {
  const [ formState, setFormState ] = useState<T>(initialValue)
  
  const onInputChange = ({target}:ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const {name, value} = target
    setFormState({
      ...formState,
      [name]: value,
    })
  }

  const onResetForm = () => {
    setFormState(initialValue)
  }

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm
  }
}
