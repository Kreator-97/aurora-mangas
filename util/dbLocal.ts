import { ShoppingCart, ShoppingItem } from '../interfaces'

export const addItemInLocal = (prevCart: ShoppingCart, newItem: ShoppingItem) => {
  const productIndex = prevCart.items.findIndex( (item) => item.product.id === newItem.product.id )
  let items: ShoppingItem[] = []

  if( productIndex === -1 ) {
    // product is not included in cart, we add it
    items = [...prevCart.items, newItem]
  } else {
    // product is included in cart, we increment the amount
    items = prevCart.items.map((item => {
      if( item.product.id === newItem.product.id ) {
        return {
          product: item.product,
          amount: item.amount + newItem.amount
        }
      }
      return item
    }))
  }

  const cart: ShoppingCart = {
    ...prevCart,
    items,
    total: (newItem.amount * newItem.product.price) + prevCart.total
  }
  updateItemsInLocal(cart)
}

export const removeItemInLocal = (prevCart: ShoppingCart, id: string) => {
  const cart: ShoppingCart = { ...prevCart, items: prevCart.items.filter( (item) => item.product.id !== id) }
  cart.total = cart.items.reduce((acc, item) => {
    return (item.product.price * item.amount + acc)
  }, 0)
  updateItemsInLocal(cart)
}

export const setProductAmountInLocal = (prevCart: ShoppingCart, amount: number, id: string ) => {
  const cart: ShoppingCart = {...prevCart, items: prevCart.items.map((item) =>( {...item, amount: item.product.id === id ? amount : item.amount })) }
  cart.total = cart.items.reduce((acc, item) => {
    return (item.product.price * item.amount + acc)
  }, 0)

  updateItemsInLocal(cart)
}

export const loadShoppingCart = ():ShoppingCart | null => {
  return JSON.parse(localStorage.getItem('cart') || 'null')
}

export const cleanShoppingCart = () => {
  const cart:ShoppingCart = {
    items: [],
    total: 0
  }

  updateItemsInLocal(cart)
}

const updateItemsInLocal = (cart: ShoppingCart ) => {
  localStorage.setItem('cart', JSON.stringify(cart))
}
