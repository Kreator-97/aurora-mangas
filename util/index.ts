export * from './formatPrice'
export * from './suggestSlug'
export * as dbLocal from './dbLocal'
export * from './formValidations'
