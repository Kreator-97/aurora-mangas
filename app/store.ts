import { configureStore } from '@reduxjs/toolkit'
import { uiSlice, authSlice, shoppingCartSlice } from './slices'

export const store = configureStore({
  reducer: {
    ui: uiSlice,
    auth: authSlice,
    cart: shoppingCartSlice
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
