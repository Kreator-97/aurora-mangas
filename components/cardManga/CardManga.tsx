import { FC } from 'react'
import Image from 'next/image'
import Link from 'next/link'
import { useRouter } from 'next/router'
import { Toaster, toast } from 'react-hot-toast'

import { addItem } from '../../app/slices/shoppingCartSlice'
import { dbLocal, formatPrice } from '../../util'
import { Manga } from '../../interfaces'
import { useAppDispatch, useAppSelector } from '../../app/hooks'

interface Props {
  manga: Manga;
}

export const CardManga:FC<Props> = ({manga}) => {
  const router = useRouter()
  const { cart } = useAppSelector(state => state)
  const dispatch = useAppDispatch()

  const onAddToCart = () => {
    dispatch( addItem({
      amount: 1,
      product: manga,
    }))

    dbLocal.addItemInLocal(cart, {amount: 1, product: manga})
    toast.success('Producto agregado al carrito')
  }

  return (
    <div className="bg-white relative min-w-[225px] max-w-[450px] rounded w-full" data-testid="card-manga">
      <Image
        src={manga.imgURL}
        layout="responsive"
        width={ 170 }
        height={ 255 }
        className="object-contain cursor-pointer rounded-t"
        onClick={ () => router.push(`/serie/${manga.serie.slug}/${manga.number}`) }
        alt={ manga.title }
      />
      <div className='p-2 flex flex-col items-center'>
        <Link passHref href={`/serie/${manga.serie.slug}`}>
          <a className='text-dark text-xl text-center hover:underline'>{ manga.serie.name } #{manga.number}</a>
        </Link>
        <p className='text-dark text-lg'>{ formatPrice(manga.price) }</p>
        <button
          className='btn bg-accent text-dark w-full'
          onClick={ () => onAddToCart()}
          role="button"
          disabled={ manga.stock === 0 }
        >
          { manga.stock === 0 ? 'Sin stock disponible' : 'Agregar al carrito' }
        </button>
      </div>
      <div className='absolute bg-dark bg-opacity-80 top-2/4 right-2 rounded-full'>
        <p className='text-2xl px-2 py-4 w-16 text-center'># {manga.number}</p>
      </div>
      <Toaster position='bottom-left' />
    </div>
  )
}
