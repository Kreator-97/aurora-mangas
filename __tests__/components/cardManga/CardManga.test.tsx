import { fireEvent, render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'

import { CardManga } from '../../../components/cardManga/CardManga'
import { mangas } from '../../fixtures/db'
import { store } from '../../../app/store'

// mocks
import { useAppDispatch, useAppSelector } from '../../../app/hooks'
import { formatPrice } from '../../../util'

jest.mock('../../../app/hooks')
jest.mock('../../../util')
jest.mock('react-hot-toast')

const dispatch = jest.fn()

const manga = mangas[0];

(useAppSelector as jest.Mock).mockReturnValue(() => ({items: [], total: 0}));
(useAppDispatch as jest.Mock).mockReturnValue(dispatch);
(formatPrice as jest.Mock).mockReturnValue('$99.00')


describe('tests on CardManga component', () => {

  test('should to match with snapshot', () => {
    const component = render(
      <Provider store={store}>
        <CardManga manga={ manga } />
      </Provider>
    )

    expect(component.container).toMatchSnapshot()
    expect(screen.getByTestId('card-manga')).toBeInTheDocument
    expect(screen.getByText(`${manga.serie.name} #${manga.number}`))
  })

  test('should to call onAddToCart event', () => {
    render(
      <Provider store={store}>
        <CardManga manga={ manga } />
      </Provider>
    )

    const button = screen.getByText('Agregar al carrito')

    expect(button).toBeInTheDocument
    fireEvent.click( button )
    expect(dispatch).toHaveBeenCalled()
    expect(dispatch).toHaveBeenCalledWith({
      payload: {
        amount: expect.any(Number),
        product: manga,
      },
      type: expect.any(String),
    })
  })
})
