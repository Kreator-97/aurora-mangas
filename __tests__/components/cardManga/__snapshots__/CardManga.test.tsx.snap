// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on CardManga component should to match with snapshot 1`] = `
<div>
  <div
    class="bg-white relative min-w-[225px] max-w-[450px] rounded w-full"
    data-testid="card-manga"
  >
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 150% 0px 0px 0px;"
      />
      <img
        class="object-contain cursor-pointer rounded-t"
        data-nimg="responsive"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
    <div
      class="p-2 flex flex-col items-center"
    >
      <a
        class="text-dark text-xl text-center hover:underline"
        href="/serie/one-punch-man"
      >
        One Punch Man
         #
        12
      </a>
      <p
        class="text-dark text-lg"
      >
        $99.00
      </p>
      <button
        class="btn bg-accent text-dark w-full"
        role="button"
      >
        Agregar al carrito
      </button>
    </div>
    <div
      class="absolute bg-dark bg-opacity-80 top-2/4 right-2 rounded-full"
    >
      <p
        class="text-2xl px-2 py-4 w-16 text-center"
      >
        # 
        12
      </p>
    </div>
  </div>
</div>
`;
